src/libponyc/expr/operator.c

```diff
@@ -373,7 +374,16 @@ bool expr_assign(pass_opt_t* opt, ast_t* ast)
 
     default: {}
   }
-  ast_t* wl_type = consume_type(fl_type, TK_NONE, false);
+
+  ast_t* wl_type;
+  if(check_auto_recover_newref(right))
+  {
+    wl_type = unisolated(fl_type);
+  }
```

Comment:
Similar comment to above, the `r_type` should be upgraded instead, as with recover blocks.